### PR TITLE
Add search for NEARESTMV mode

### PR DESCRIPTION
**Pull request by fbossen**
Move MV stack construction out of encoding function such that the
value of the nearest MV can be determined before encoding

### CODE REVIEW

**Comment by fbossen** [comment]
|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
| -6.6882 | -7.3289 | -7.6921 |  -6.5564 | -6.5062 | -6.5456 |    -6.8485 |

**Comment by ycho** [comment]
@atomnuker pls reviewer this patch

File: src/context.rs [hunk 1]
```diff
@@ -2396,19 +2404,21 @@ impl ContextWriter {
...
         let found_match = self.scan_col_mbmi(bo, col_offset, max_col_offs, &mut processed_cols, ref_frame, mv_stack,
-                                             &mut far_newmv_count);
+                                             &mut far_newmv_count, bsize);
         col_match |= found_match;
       }
     }
 
     let total_match = if row_match { 1 } else { 0 } + if col_match { 1 } else { 0 };
 
+    assert!(total_match >= nearest_match);
```
Comment (atomnuker) [inline src/context.rs]: Why? Doesn't seem like it'll be possible unless someone unintentionally breaks the code (at which point a desync should happen).

File: src/context.rs [hunk 2]
```diff
@@ -2131,27 +2134,28 @@ impl ContextWriter {
...
      * to the right won't have been decoded) */
-    if (blk.n4_w > blk.n4_h) && blk.is_sec_rect {
+    if (target_n4_w > target_n4_h) && is_sec_rect {
       has_tr = false;
     }
 
     /* The bottom left square of a Vertical A (in the old format) does
      * not have a top right as it is decoded before the right hand
      * rectangle of the partition */
+/*
```
Comment (atomnuker) [inline src/context.rs]: Why disable this? Isn't it required to exclude some edge cases?

File: src/context.rs [hunk 3]
```diff
@@ -2428,7 +2438,7 @@ impl ContextWriter {
   }
 
   pub fn find_mvrefs(&mut self, bo: &BlockOffset, ref_frame: usize,
-                     mv_stack: &mut Vec<CandidateMV>) -> usize {
+                     mv_stack: &mut Vec<CandidateMV>, bsize: BlockSize, is_sec_rect: bool) -> usize {
```
Comment (atomnuker) [inline src/context.rs]: Why did you change from using blk.is_sec_rect to needing it as a parameter? Its always set to false now when you call find_mvrefs.

File: src/context.rs (hunk 1 above)
Comment (fbossen) [reply src/context.rs]: Yes, this won't ever happen. The only reason it's there is to facilitate understanding of the lines below. In particular, the fact that total_match is equal to at least 1 when nearest_match is 1 may not be obvious. Alternatively this could me put in a comment

File: src/context.rs (hunk 2 above)
Comment (fbossen) [reply src/context.rs]: I commented it out for now to avoid relying on blk (current block). Down the road we'll have to decide whether to write things into blk before calling find_mvrefs() or pass values as parameters. For now I am passing values as parameters.

File: src/context.rs (hunk 3 above)
Comment (fbossen) [reply src/context.rs]: blk.is_sec_rect would always be false for now. Same issue as in previous comment: do we pass values as parameters or do we write them into the current blk structure and read them out?

File: src/rdo.rs [hunk 4]
```diff
@@ -240,10 +243,10 @@ pub fn rdo_mode_decision(
     }
 
     let ref_frame = if luma_mode.is_intra() { INTRA_FRAME } else { LAST_FRAME };
-    let mv = if luma_mode != PredictionMode::NEWMV {
-      MotionVector { row: 0, col: 0 }
-    } else {
-      motion_estimation(fi, fs, bsize, bo, ref_frame)
+    let mv = match luma_mode {
```
Comment (ycho) [inline src/rdo.rs]: Once mv is decided here, is "best_mv = mv;" at line 293 still required?

File: src/rdo.rs [hunk 5]
```diff
@@ -297,6 +301,7 @@ pub fn rdo_mode_decision(
   }
 
   cw.bc.set_mode(bo, bsize, best_mode_luma);
+  cw.bc.set_motion_vector(bo, bsize, best_mv);
```
Comment (ycho) [inline src/rdo.rs]: Also, unless best_mv is updated, we can do this right after "let mv = ...." ?

File: src/context.rs [hunk 6]
```diff
@@ -2321,14 +2326,18 @@ impl ContextWriter {
     }
   }
 
-  fn setup_mvref_list(&mut self, bo: &BlockOffset, ref_frame: usize, mv_stack: &mut Vec<CandidateMV>) -> usize {
+  fn setup_mvref_list(&mut self, bo: &BlockOffset, ref_frame: usize, mv_stack: &mut Vec<CandidateMV>,
+                      bsize: BlockSize, is_sec_rect: bool) -> usize {
     let (_rf, _rf_num) = self.get_mvref_ref_frames(INTRA_FRAME);
 
+    let target_n4_h = bsize.height_mi();
+    let target_n4_w = bsize.height_mi();
```
[the remark this answers is not in the document]
Comment (fbossen) [reply src/context.rs]: Nice catch! Will fix

File: src/rdo.rs (hunk 5 above)
Comment (fbossen) [reply src/rdo.rs]: No. Basically this is needed so that the RDO for the next block behaves correctly.

File: src/rdo.rs (hunk 4 above)
Comment (fbossen) [reply src/rdo.rs]: Yes, my understanding is that the best_mv needs to be captured in the RDOOutput so that it can be used in the final encode.

Comment (ycho) [reply src/rdo.rs]: So, initially  best_mv = MotionVector { row: 0, col: 0 };, then assume intra mode is better than mv modes and no chance for doing "best_mv = mv;", then best_mv is still (0,0). Would this be expected?

**Comment by fbossen** [comment]
Seems even better on a recent version of master:

|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
| -7.7694 | -7.9452 | -8.2971 |  -7.7014 | -7.7181 | -7.9676 |    -7.7985 |
